src/frontend/Typechecker.ml

```diff
@@ -1028,13 +1028,60 @@ let check_assignment_operator loc assop lhs rhs =
       let return_type = assignmentoperator_stan_math_return_type op args in
       match return_type with Some Void -> rhs | _ -> err op |> error )
 
-let rec check_lvalue cf tenv = function
-  | {lval= LVariable id; lmeta= ({loc} : located_meta)} ->
+(** When an lvalue is assigning to multiple places at once (e.g., a tuple
+    is being unpacked), we want to ensure that the same memory is not being
+    written to multiple times. This lets us preserve the illusion of
+    simultaneous assignment and generally avoid confusion. *)
+let verify_lvalue_unique lv =
+  let rec flatten_lvalues lv =
+    match (lv.lval, lv.lmeta.type_) with
+    | LTuplePacking lvs, _ -> List.concat_map ~f:flatten_lvalues lvs
+    | LIndexed (l, _), _ ->
+        (* Prevent assigning to multiple indices in the same object at once.
+           In principal it is safe to assign to disjoint indices, but statically
+           checking that is impossible. *)
+        flatten_lvalues l
+    | _, UTuple ts ->
+        (* If we're assigning an entire tuple, we also need to prevent
+           assigning to any slot in this statement. *)
+        lv
+        :: List.concat_mapi ts ~f:(fun i ty ->
+               flatten_lvalues
+                 { lval= LTupleProjection (lv, i + 1)
+                 ; lmeta= {lv.lmeta with type_= ty} } )
+    | _ -> [lv] in
+  let all_lvals =
+    flatten_lvalues lv |> List.map ~f:Ast.untyped_lvalue_of_typed_lvalue in
+  match List.find_all_dups ~compare:Ast.compare_untyped_lval all_lvals with
```

Comment:
`compare_untyped_lval` is not correct here, e.g.
```stan
int i = 1, j = 1;
array[1] tuple(int, int) tup;
(tup[i].1, tup[j].1) = (1, 2);
(tup, tup[1].1) = ({(1,2)}, 3);
```
I'd erase all array-indexing when flattening the lvalues
```ocaml
  let rec flatten idx lv = function
    match lv.lval with
    | LTuplePacking lvs -> List.concat_map ~f:(flatten []) lvs
    | LIndexed (l,_) -> flatten idx lv
    | LTupleProjection (l, ix) -> flatten (ix :: idx) l
    | LVariable id -> [(id.name, idx)]
```

Reply:
These were tricky but I think it is working now.

One unfortunate result of stripping away this information is we can't then print it, so we have to resort to just printing `tup` rather than `tup.1` (misleading) or `tup[].1` (ideal, but that information is lost) in the errors

Reply:
Ok, I found a reasonable way of having the error mention `tup[...].1`